### PR TITLE
Add TCP+TLS Healthchecks

### PR DESCRIPTION
**Pull request by pgporada**
### Description

Allows TCP healthchecks to optionally connect to a listening TCP server over TLS.

Fixes https://github.com/hashicorp/consul/issues/3936.

### PR Checklist

* [X] updated test coverage
* [X] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern - well, sort of?

### CODE REVIEW

File: tlsutil/config.go [hunk 1]
```diff
@@ -878,6 +878,7 @@ func (c *Configurator) OutgoingTLSConfigForCheck(skipVerify bool, serverName str
 	c.log("OutgoingTLSConfigForCheck")
 
 	c.lock.RLock()
+
```
Comment (beautifulentropy) [inline tlsutil/config.go:881]:
```suggestion
```

File: website/content/docs/ecs/configuration-reference.mdx [hunk 2]
```diff
@@ -189,6 +189,7 @@ Defines the Consul checks for the service. Each `check` object may contain the f
 | `status` | `string` | optional | Specifies the initial status the health check. Must be one of `passing`, `warning`, `critical`, `maintenance`, or `null`. |
 | `successBeforePassing` | `integer` | optional | Specifies the number of consecutive successful results required before check status transitions to passing.  |
 | `tcp` | `string` | optional | Specifies this is a TCP check. Must be an IP/hostname plus port to which a TCP connection is made every `interval`.  |
+| `tcpUseTls` | `boolean` | optional | Specifies whether to use TLS for this TCP health check.  |
 | `timeout` | `string` | optional | Specifies a timeout for outgoing connections. Applies to script, HTTP, TCP, UDP, and gRPC checks. Must be a duration string, such as `10s` or `5m`.  |
 | `tlsServerName` | `string` | optional | Specifies an optional string used to set the SNI host when connecting via TLS.  |
 | `tlsSkipVerify` | `boolean` | optional | Specifies if the certificate for an HTTPS check should not be verified.  |
```
Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]:
```suggestion
| `tlsSkipVerify` | `boolean` | optional | Determines if the check verifies the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]
Looks great, just some small nits and some additional documentation that needs updating.


Update the the documentation, here: `website/content/docs/services/configuration/checks-configuration-reference.mdx`

Modify line `39`:
```markdown
| `tls_skip_verify` | Boolean value that determines if the check verifies the chain and hostname of the certificate that the server presents. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> <li>TCP </li> |
```

Add line below `46` (under `tcp:`):
```markdown
| `tcp_use_tls` | Boolean value that enables TLS for TCP checks when set to `true`. | <li>gRPC </li> |
```

**Comment by pgporada** [comment]
Thanks for the review, all the doc changes have been completed.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 3]
```diff
@@ -26,30 +26,30 @@ Specify health check options in the `check` block. To register two or more heath
...
 | `os_service` | String value that specifies the name of the name of a service to check during an OSService check. | <li>OSService</li> |
 | `service_id` | String value that specifies the ID of a service instance to associate with an OSService check. That service instance must be on the same node as the check. If not specified, the check verifies the health of the node. | <li>OSService</li> |
 | `tcp` | String value that specifies an IP address or host and port number for the check establish a TCP connection with. | <li>TCP</li> |
+| `tcp_use_tls` | Boolean value that enables TLS for TCP checks when set to `true`. | <li>TCP </li> |
 | `udp` | String value that specifies an IP address or host and port number for the check to send UDP datagrams to. | <li>UDP</li> |
 | `ttl` | String value that specifies how long to wait for an update from an external process during a TTL check. | <li>TTL</li> |
-| `alias_service` | String value that specifies a service or node that the service associated with the health check aliases. | <li>Alias</li> | 
+| `alias_service` | String value that specifies a service or node that the service associated with the health check aliases. | <li>Alias</li> |
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:50]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 4]
```diff
@@ -26,30 +26,30 @@ Specify health check options in the `check` block. To register two or more heath
...
+| `tls_skip_verify` | Boolean value that determines if the check verifies the chain and hostname of the certificate that the server presents. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> <li>TCP </li> |
 | `method` | String value that specifies the request method to send during HTTP checks. Default is `GET`. | <li>HTTP</li> |
 | `header` | Object that specifies header fields to send in HTTP check requests. Each header specified in `header` object contains a list of string values. | <li>HTTP</li> |
 | `body` | String value that contains JSON attributes to send in HTTP check requests. You must escap the quotation marks around the keys and values for each attribute. | <li>HTTP</li> |
-| `disable_redirects` | Boolean value that prevents HTTP checks from following redirects if set to `true`. Default is `false`. | <li>HTTP</li> |  
+| `disable_redirects` | Boolean value that prevents HTTP checks from following redirects if set to `true`. Default is `false`. | <li>HTTP</li> |
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:43]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 5]
```diff
@@ -26,30 +26,30 @@ Specify health check options in the `check` block. To register two or more heath
...
-| `grpc` | String value that specifies the gRPC endpoint, including port number, to send requests to. Append the endpoint with `:/` and a service identifier to check a specific service. The endpoint must support the [gRPC health checking protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md). | <li>gRPC </li>|        
+| `grpc` | String value that specifies the gRPC endpoint, including port number, to send requests to. Append the endpoint with `:/` and a service identifier to check a specific service. The endpoint must support the [gRPC health checking protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md). | <li>gRPC </li>|
 | `grpc_use_tls` | Boolean value that enables TLS for gRPC checks when set to `true`. | <li>gRPC </li> |
-| `h2ping` | String value that specifies the HTTP2 endpoint, including port number, to send HTTP2 requests to. | <li>H2ping</li> |        
+| `h2ping` | String value that specifies the HTTP2 endpoint, including port number, to send HTTP2 requests to. | <li>H2ping</li> |
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:35]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 6]
```diff
@@ -26,30 +26,30 @@ Specify health check options in the `check` block. To register two or more heath
...
 | `docker_container_id` | Specifies the Docker container ID in which to run an external health check application. Specify the external application with the `args` parameter. | <li> Docker </li>  |
 | `shell` | String value that specifies the type of command line shell to use for running the health check application. Specify the external application with the `args` parameter. | <li> Docker </li>  |
-| `grpc` | String value that specifies the gRPC endpoint, including port number, to send requests to. Append the endpoint with `:/` and a service identifier to check a specific service. The endpoint must support the [gRPC health checking protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md). | <li>gRPC </li>|        
+| `grpc` | String value that specifies the gRPC endpoint, including port number, to send requests to. Append the endpoint with `:/` and a service identifier to check a specific service. The endpoint must support the [gRPC health checking protocol](https://github.com/grpc/grpc/blob/master/doc/health-checking.md). | <li>gRPC </li>|
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:33]: Revert this line change.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 7]
```diff
@@ -13,10 +13,10 @@ This topic provides configuration reference information for health checks. For i
 Health checks perform several safety functions, such as allowing a web balancer to gracefully remove failing nodes and allowing a database to replace a failed secondary. You can configure health checks to monitor the health of the entire node. Refer to [Define Health Checks](/consul/docs/services/usage/checks) for information about how to define the differnet types of health checks available in Consul.
 
 ## Check block
-Specify health check options in the `check` block. To register two or more heath checks in the same configuration, use the [`checks` block](#checks-block). The following table describes the configuration options contained in the `check` block. 
+Specify health check options in the `check` block. To register two or more heath checks in the same configuration, use the [`checks` block](#checks-block). The following table describes the configuration options contained in the `check` block.
 
-| Parameter | Description | Check types | 
-| ---       | ---          | ---         | 
+| Parameter | Description | Check types |
+| ---       | ---          | ---         |
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:19]: Revert both of these line changes.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 8]
```diff
@@ -13,10 +13,10 @@ This topic provides configuration reference information for health checks. For i
 Health checks perform several safety functions, such as allowing a web balancer to gracefully remove failing nodes and allowing a database to replace a failed secondary. You can configure health checks to monitor the health of the entire node. Refer to [Define Health Checks](/consul/docs/services/usage/checks) for information about how to define the differnet types of health checks available in Consul.
 
 ## Check block
-Specify health check options in the `check` block. To register two or more heath checks in the same configuration, use the [`checks` block](#checks-block). The following table describes the configuration options contained in the `check` block. 
+Specify health check options in the `check` block. To register two or more heath checks in the same configuration, use the [`checks` block](#checks-block). The following table describes the configuration options contained in the `check` block.
```
Comment (beautifulentropy) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:16]: Revert this line change.

File: website/content/docs/ecs/configuration-reference.mdx [hunk 9]
```diff
@@ -189,9 +189,10 @@ Defines the Consul checks for the service. Each `check` object may contain the f
...
 | `tcp` | `string` | optional | Specifies this is a TCP check. Must be an IP/hostname plus port to which a TCP connection is made every `interval`.  |
+| `tcpUseTls` | `boolean` | optional | Specifies whether to use TLS for this `TCP` health check. If TLS is enabled, then by default, a valid TLS certificate is expected. Certificate verification can be disabled by setting `TLSSkipVerify` to `true`.  |
 | `timeout` | `string` | optional | Specifies a timeout for outgoing connections. Applies to script, HTTP, TCP, UDP, and gRPC checks. Must be a duration string, such as `10s` or `5m`.  |
 | `tlsServerName` | `string` | optional | Specifies an optional string used to set the SNI host when connecting via TLS.  |
-| `tlsSkipVerify` | `boolean` | optional | Specifies if the certificate for an HTTPS check should not be verified.  |
+| `tlsSkipVerify` | `boolean` | optional | Determines if the check verifies the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```
Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]: My bad, use verbiage that's more consistent with the other lines in this file: 

```suggestion
| `tlsSkipVerify` | `boolean` | optional | Specifies if the check should verify the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

Comment (beautifulentropy) [inline website/content/docs/ecs/configuration-reference.mdx:195]: My bad, use verbiage that's more consistent with the other lines in this file: 

```suggestion
| `tlsSkipVerify` | `boolean` | optional | Specifies if the check should verify the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]

File: agent/agent_test.go [hunk 10]
```diff
@@ -967,6 +970,135 @@ func TestAgent_AddServiceWithH2CPINGCheck(t *testing.T) {
...
+// be stopped by calling Close().
+func newTestTCPServer(t *testing.T) *testTCPServer {
+	// Yes I get it, we're passing the same cert from the Consul Agent as is
+	// configured in this toy application server. Don't ever do this except in
+	// testing. I see you, thinking about it, don't.
+	certFile := "../test/key/ourdomain_server.cer"
+	keyFile := "../test/key/ourdomain_server.key"
+	caFile := "../test/ca/root.cer"
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	require.Equal(t, err, nil)
```
Comment (beautifulentropy) [inline agent/agent_test.go:991]: Here and elsewhere, we should use `require.NoError()`.

```suggestion
	require.NoError(t, err)
```

File: agent/agent_test.go [hunk 11]
```diff
@@ -967,6 +970,135 @@ func TestAgent_AddServiceWithH2CPINGCheck(t *testing.T) {
...
+	// Configure TLS to require and verify the client's certificate
+	config := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		ClientCAs:    rootCertPool,
+	}
+
+	l, err := tls.Listen("tcp", "127.0.0.1:0", config)
+	require.Equal(t, err, nil)
+	log.Printf("test tcp server listening on %s", l.Addr())
```
Comment (beautifulentropy) [inline agent/agent_test.go:1009]: Here and elsewhere, we should drop these log lines.

```suggestion
```

File: agent/agent_test.go [hunk 12]
```diff
@@ -967,6 +970,135 @@ func TestAgent_AddServiceWithH2CPINGCheck(t *testing.T) {
...
+
+	// Retry until the healthcheck is passing.
+	retry.Run(t, func(r *retry.R) {
+		status := getCheck(a, "arbitraryTCPServerTLSCheck")
+		if status.Status != api.HealthPassing {
+			r.Fatalf("bad: %v", status.Status)
+		}
+	})
+}
+
```
Comment (beautifulentropy) [inline agent/agent_test.go:1101]: The implementation you have here is great for debugging your implementation. However, for the final test case, we should go with something smaller that uses more of the standard library:

```suggestion
func startMockTLSServer(t *testing.T) (addr string, closeFunc func()) {
	// Load certificates
	cert, err := tls.LoadX509KeyPair("../test/key/ourdomain_server.cer", "../test/key/ourdomain_server.key")
	require.NoError(t, err)

	// Create a certificate pool
	rootCertPool := x509.NewCertPool()
	caCert, err := os.ReadFile("../test/ca/root.cer")
	require.NoError(t, err)
	rootCertPool.AppendCertsFromPEM(caCert)

	// Configure TLS
	config := &tls.Config{
		Certificates: []tls.Certificate{cert},
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    rootCertPool,
	}

	// Start TLS server
	ln, err := tls.Listen("tcp", "127.0.0.1:0", config)
	require.NoError(t, err)

	go func() {
		for {
			conn, err := ln.Accept()
			if err != nil {
				return
			}
			go func(c net.Conn) {
				io.Copy(c, c)
				c.Close()
			}(conn)
		}
	}()

	return ln.Addr().String(), func() { ln.Close() }
}

func TestAgent_AddServiceWithTCPTLSCheck(t *testing.T) {
	t.Parallel()
	dataDir := testutil.TempDir(t, "agent")
	a := NewTestAgent(t, `
		data_dir = "`+dataDir+`"
		enable_agent_tls_for_checks = true
		datacenter = "dc1"
		tls {
			defaults {
				ca_file   = "../test/ca/root.cer"
				cert_file = "../test/key/ourdomain_server.cer"
				key_file  = "../test/key/ourdomain_server.key"
			}
		}
	`)
	defer a.Shutdown()
	testrpc.WaitForTestAgent(t, a.RPC, "dc1")

	// Start mock TCP+TLS server
	addr, closeServer := startMockTLSServer(t)
	defer closeServer()

	check := &structs.HealthCheck{
		Node:    "foo",
		CheckID: "arbitraryTCPServerTLSCheck",
		Name:    "arbitraryTCPServerTLSCheck",
		Status:  api.HealthCritical,
	}

	chkType := &structs.CheckType{
		TCP:           addr,
		TCPUseTLS:     true,
		TLSServerName: "server.dc1.consul",
		Interval:      5 * time.Second,
	}

	err := a.AddCheck(check, chkType, false, "", ConfigSourceLocal)
	require.NoError(t, err)

	// Retry until the healthcheck is passing.
	retry.Run(t, func(r *retry.R) {
		status := getCheck(a, "arbitraryTCPServerTLSCheck")
		if status.Status != api.HealthPassing {
			r.Fatalf("bad: %v", status.Status)
		}
	})
}

```

**Review by beautifulentropy — CHANGES_REQUESTED** [review]

**Comment by trujillo-adam** [comment]
I don't see any updates to the main configuration reference page for health checks: https://developer.hashicorp.com/consul/docs/services/configuration/checks-configuration-reference

Are these updates not relevant?

**Comment by pgporada** [comment]
That page is relevant, it was forgotten in the mix. Apologies.

**Comment by pgporada** [comment]
@trujillo-adam Turns out that the file that builds the docs at the link you sent was already in my PR. See https://github.com/hashicorp/consul/pull/18381/files#diff-c4d30ce7367e53ab22681c3938502abd8fefd01546c6c85a4454c21b45bd7100R47.

File: agent/checks/check.go [hunk 13]
```diff
@@ -694,17 +695,27 @@ func (c *CheckTCP) run() {
 
 // check is invoked periodically to perform the TCP check
 func (c *CheckTCP) check() {
-	conn, err := c.dialer.Dial(`tcp`, c.TCP)
```
Comment (loshz) [inline agent/checks/check.go:697]: I think we can simplify this whole function if we use an interface for the conn:
```go
func (c *CheckTCP) check() {
	var conn io.Closer
	var err error

	if c.TLSClientConfig == nil {
		conn, err = c.dialer.Dial(`tcp`, c.TCP)
	} else {
		conn, err = tls.DialWithDialer(c.dialer, `tcp`, c.TCP, c.TLSClientConfig)
	}

	if err != nil {
		c.Logger.Warn("Check TCP connection failed",
			"check", c.CheckID.String(),
			"error", err,
		)
		c.StatusHandler.updateCheck(c.CheckID, api.HealthCritical, err.Error())
		return
	}

	conn.Close()
	c.StatusHandler.updateCheck(c.CheckID, api.HealthPassing, fmt.Sprintf("TCP connect %s: Success", c.TCP))
}
```

WDYT?

File: website/content/docs/ecs/configuration-reference.mdx [hunk 14]
```diff
@@ -189,9 +189,10 @@ Defines the Consul checks for the service. Each `check` object may contain the f
...
 | `tcp` | `string` | optional | Specifies this is a TCP check. Must be an IP/hostname plus port to which a TCP connection is made every `interval`.  |
+| `tcpUseTls` | `boolean` | optional | Specifies whether to use TLS for this `TCP` health check. If TLS is enabled, then by default, a valid TLS certificate is expected. Certificate verification can be disabled by setting `TLSSkipVerify` to `true`.  |
 | `timeout` | `string` | optional | Specifies a timeout for outgoing connections. Applies to script, HTTP, TCP, UDP, and gRPC checks. Must be a duration string, such as `10s` or `5m`.  |
 | `tlsServerName` | `string` | optional | Specifies an optional string used to set the SNI host when connecting via TLS.  |
-| `tlsSkipVerify` | `boolean` | optional | Specifies if the certificate for an HTTPS check should not be verified.  |
+| `tlsSkipVerify` | `boolean` | optional | Specifies if the check should verify the chain and hostname of the certificate presented by the server being checked. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. Supported check types: `HTTP`, `H2Ping`, `gRPC`, and `TCP`|
```
Comment (loshz) [inline website/content/docs/ecs/configuration-reference.mdx:195]: I am ok with this new wording, but I'll will have to defer to someone from our security posture to see what they think.

File: website/content/docs/services/configuration/checks-configuration-reference.mdx [hunk 15]
```diff
@@ -36,14 +36,15 @@ Specify health check options in the `check` block. To register two or more heath
...
+| `tls_skip_verify` | Boolean value that determines if the check verifies the chain and hostname of the certificate that the server presents. Set to `true` to disable verification. We recommend setting to `false` for production use. Default is `false`. | <li>HTTP </li> <li>H2Ping </li> <li>gRPC </li> <li>TCP </li> |
 | `method` | String value that specifies the request method to send during HTTP checks. Default is `GET`. | <li>HTTP</li> |
 | `header` | Object that specifies header fields to send in HTTP check requests. Each header specified in `header` object contains a list of string values. | <li>HTTP</li> |
-| `body` | String value that contains JSON attributes to send in HTTP check requests. You must escape the quotation marks around the keys and values for each attribute. | <li>HTTP</li> |
+| `body` | String value that contains JSON attributes to send in HTTP check requests. You must escap the quotation marks around the keys and values for each attribute. | <li>HTTP</li> |
```
Comment (loshz) [inline website/content/docs/services/configuration/checks-configuration-reference.mdx:42]: Revert this change.

**Review by loshz — CHANGES_REQUESTED** [review]
Hi @pgporada - thanks for working on this, and sorry it took so long for me to review! 

This is a very useful feature that I know has been requested for a while, so I appreciate you taking the time to work on it.

I've left a couple of comments, and need to also get another eng from Hashi to validate a few things, but I can do that.

Feel free to ping me if you have any questions!

File: agent/checks/check.go (hunk 13 above)
Comment (pgporada) [reply agent/checks/check.go:697]: That's a great idea, I hadn't considered that was possible.

**Comment by pgporada** [comment]
Thank you for taking the time to review this. :)

File: agent/agent_test.go [hunk 16]
```diff
@@ -967,6 +968,80 @@ func TestAgent_AddServiceWithH2CPINGCheck(t *testing.T) {
...
+	require.NoError(t, err)
+	// Create a certificate pool
+	rootCertPool := x509.NewCertPool()
+	caCert, err := os.ReadFile("../test/ca/root.cer")
+	require.NoError(t, err)
+	rootCertPool.AppendCertsFromPEM(caCert)
+	// Configure TLS
+	config := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientAuth:   tls.RequireAndVerifyClientCert,
```
Comment (pgporada) [inline agent/agent_test.go:983]: For lines 982 and 983, I chose not to generate new key material. The `tls.RequireAndVerifyClientCert` is working by virtue of the Consul Agent sending a TLS cert with a SAN that matches a SAN loaded by the mockTLSServer. If you want me to generate new key material and change this, that's fine by me.

File: agent/agent_test.go [hunk 17]
```diff
@@ -14,11 +14,15 @@ import (
...
 	"fmt"
+<<<<<<< HEAD
+	"io"
+=======
 	"github.com/hashicorp/consul/agent/grpc-external/limiter"
 	"github.com/hashicorp/consul/agent/proxycfg"
 	"github.com/hashicorp/consul/agent/proxycfg-sources/local"
 	"github.com/hashicorp/consul/agent/xds"
 	proxytracker "github.com/hashicorp/consul/internal/mesh/proxy-tracker"
+>>>>>>> main
```
Comment (loshz) [inline agent/agent_test.go:25]: Looks like a merge conflict was missed here?

Comment (pgporada) [reply agent/agent_test.go]: Fixed, sorry about that.

**Comment by loshz** [comment]
@david-yu do we want to backport this? Or is it deemed a "new feature"?

**Comment by david-yu** [comment]
I would backport to just 1.16.x just to limit blast radius if there are issues. I think it would be good to backport to at least get usage of this feature in our latest stable branch.

**Comment by loshz** [comment]
@pgporada thanks for fixing the merge conflict and addressing the pr feedback. One last thing we need to do is add a changelog file (`.changelog/18381.txt`), similar to something like [this](https://github.com/hashicorp/consul/blob/main/.changelog/18617.txt), that outlines the specific change made here.

**Comment by pgporada** [comment]
@loshz Done, thanks for the heads up on that.

**Review by loshz — APPROVED** [review]
LGTM - I appreciate you taking the time to see this through and making the appropriate changes!

FYI, once this gets merged, a follow up PR will be generated in order to backport this to 1.16 versions. It should get auto merged, but I'll keep an eye on it in case it doesn't.

**Pull request merged**